packages/x-data-grid/src/components/containers/GridRootStyles.ts

```diff
@@ -273,13 +275,40 @@ export const GridRootStyles = styled('div', {
         t.vars
           ? `rgba(${t.vars.palette.primary.mainChannel} / 0.5)`
           : alpha(t.palette.primary.main, 0.5)
-      } 1px`,
-      outlineWidth: 1,
-      outlineOffset: -1,
+      } ${focusOutlineWidth}px`,
+      outlineOffset: focusOutlineWidth * -1,
     },
     [`& .${c.columnHeader}:focus, & .${c.cell}:focus`]: {
-      outline: `solid ${t.palette.primary.main} 1px`,
+      outline: `solid ${t.palette.primary.main} ${focusOutlineWidth}px`,
+      outlineOffset: focusOutlineWidth * -1,
+    },
+    // Hide the column separator when:
+    // - the column is focused and has an outline
+    // - the next column is focused and has an outline
+    // - the column has a right border
+    // - the next column has a right border
+    [`& .${c.columnHeader}:focus .${c.columnSeparator},
+      & .${c.columnHeader}:focus-within .${c.columnSeparator},
+      & .${c.columnHeader}:has(+ .${c.columnHeader}:focus) .${c.columnSeparator},
+      & .${c.columnHeader}:has(+ .${c.columnHeader}:focus-within) .${c.columnSeparator},
+      & .${c['columnHeader--withRightBorder']} .${c.columnSeparator},
+      & .${c.columnHeader}:has(+ .${c['columnHeader--withRightBorder']}) .${c.columnSeparator}`]: {
+      opacity: 0,
+      '@media (hover: none)': {
+        opacity: 1,
+        color: (t.vars || t).palette.primary.main,
+      },
     },
+    // Show resizable separators again when the column is hovered
+    [`& .${c.columnHeader}:focus .${c['columnSeparator--resizable']}:hover,
+      & .${c.columnHeader}:focus-within .${c['columnSeparator--resizable']}:hover,
+      & .${c.columnHeader}:has(+ .${c.columnHeader}:focus) .${c['columnSeparator--resizable']}:hover,
+      & .${c.columnHeader}:has(+ .${c.columnHeader}:focus-within) .${c['columnSeparator--resizable']}:hover,
+      & .${c['columnHeader--withRightBorder']} .${c['columnSeparator--resizable']}:hover,
+      & .${c.columnHeader}:has(+ .${c['columnHeader--withRightBorder']}) .${c['columnSeparator--resizable']}:hover`]:
+      {
+        opacity: 1,
```

Comment:
I think this can be more compact:

```tsx
[`& .${c.columnHeader}:focus,
  & .${c.columnHeader}:focus-within,
  & .${c.columnHeader}:has(+ .${c.columnHeader}:focus),
  & .${c.columnHeader}:has(+ .${c.columnHeader}:focus-within),
  & .${c['columnHeader--withRightBorder']},
  & .${c.columnHeader}:has(+ .${c['columnHeader--withRightBorder']})`]: {
  [`& .${c.columnSeparator}`]: {
    opacity: 0,
    '@media (hover: none)': {
      opacity: 1,
      color: (t.vars || t).palette.primary.main,
    },
  },
  // Show resizable separators again when the column is hovered
  [`& .${c['columnSeparator--resizable']}:hover`]: {
    opacity: 1,
  },
},
```

Not sure if it's a good idea though, feel free to disregard if you prefer.

Reply:
Less repetitive, I like it.